### PR TITLE
opaque type cleanup

### PR DESCRIPTION
**Pull request by lcnr**
the commits are pretty self-contained.

r? @compiler-errors cc @oli-obk

### CODE REVIEW

File: compiler/rustc_trait_selection/src/solve/eval_ctxt/canonical.rs [hunk 1]
```diff
@@ -317,12 +316,8 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
...
         opaque_types: &[(ty::OpaqueTypeKey<'tcx>, Ty<'tcx>)],
     ) -> Result<(), NoSolution> {
-        for &(a, b) in opaque_types {
-            let InferOk { value: (), obligations } =
-                self.infcx.register_hidden_type_in_new_solver(a, param_env, b)?;
-            // It's sound to drop these obligations, since the normalizes-to goal
-            // is responsible for proving these obligations.
-            let _ = obligations;
+        for &(key, ty) in opaque_types {
+            self.insert_hidden_type(key, param_env, ty)?;
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/solve/eval_ctxt/canonical.rs:320]: this was previously broken. `register_hidden_type` returns both obligations from checking the item bounds and from equating the added hidden type with the previous one. Can't be bothered to write a test for that though :sweat_smile: especially cause wf of opaque types is checked separately anyways ^^

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/solve/eval_ctxt/canonical.rs:320]: oof

**Review by compiler-errors — APPROVED** [review]

File: compiler/rustc_borrowck/src/region_infer/opaque_types.rs [hunk 2]
```diff
@@ -347,10 +342,10 @@ impl<'tcx> InferCtxtExt<'tcx> for InferCtxt<'tcx> {
 fn check_opaque_type_parameter_valid(
     tcx: TyCtxt<'_>,
     opaque_type_key: OpaqueTypeKey<'_>,
-    origin: OpaqueTyOrigin,
     span: Span,
 ) -> Result<(), ErrorGuaranteed> {
-    match origin {
+    let opaque_ty_hir = tcx.hir().expect_item(opaque_type_key.def_id);
+    match opaque_ty_hir.expect_opaque_ty().origin {
```
Comment (oli-obk) [inline compiler/rustc_borrowck/src/region_infer/opaque_types.rs:357]: We also have the `is_type_alias_impl_trait` query

**Pull request merged**